collector/devstat_freebsd.go

```diff
@@ -0,0 +1,239 @@
+	stats.duration.free = duration_free;
+	stats.busyTime = busy_time;
+	stats.blocks = blocks;
+
+	return stats;
+}
+*/
+import "C"
+
+const (
+	devstatSubsystem = "devstat"
+)
+
+type devstatCollector struct {
+	bytes     *prometheus.GaugeVec
+	transfers *prometheus.GaugeVec
+	duration  *prometheus.GaugeVec
+	busyTime  *prometheus.GaugeVec
+	blocks    *prometheus.GaugeVec
+}
+
+func init() {
+	Factories["devstat"] = NewDevstatCollector
+}
+
+// device stats.
+func NewDevstatCollector() (Collector, error) {
+	//var diskLabelNames = []string{"device"}
+	//var ioType = []string{"type"}
+
+	return &devstatCollector{
+		// Docs from https://www.kernel.org/doc/Documentation/iostats.txt
+		bytes: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Subsystem: devstatSubsystem,
+				Name:      "bytes",
+				Help:      "The total number of bytes transferred",
+			},
+			[]string{"device", "type"},
+		),
+		transfers: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Subsystem: devstatSubsystem,
+				Name:      "transfers",
+				Help:      "The total number of transactions",
+			},
+			[]string{"device", "type"},
+		),
+		duration: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Subsystem: devstatSubsystem,
+				Name:      "duration",
+				Help:      "The total duration of transactions",
+			},
+			[]string{"device", "type"},
+		),
+		busyTime: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Subsystem: devstatSubsystem,
+				Name:      "busy_time",
+				Help:      "Total time the device had one or more transactions outstanding",
+			},
+			[]string{"device"},
+		),
+		blocks: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: Namespace,
+				Subsystem: devstatSubsystem,
+				Name:      "blocks",
+				Help:      "The total number of blocks transferred",
+			},
+			[]string{"device"},
+		),
+	}, nil
+}
+
+func (c *devstatCollector) Update(ch chan<- prometheus.Metric) (err error) {
+	/*
+		var busyTime [3]C.longlong
+		var blocks [3]C.uint64_t
+		var kbPerTransfer [3]C.longlong
+		var transfersPerSecond [3]C.longlong
+		var mbPerSecond [3]C.longlong
+		var blocksPerSecond [3]C.longlong
+		var msPerTransaction [3]C.longlong
+		var busyPCT C.longlong
+		var queueLength C.uint64_t
+
+	*/
+	count := C._get_ndevs()
+	if count == -1 {
+		return errors.New("devstat_getdevs() failed!")
+	}
+	if count == -2 {
+		return errors.New("calloc() failed!")
+	}
+
+	for i := C.int(0); i < count; i++ {
+		stats := C._get_stats(i)
+		device := fmt.Sprintf("%s%d", C.GoString(&stats.device[0]), stats.unit)
+		c.bytes.With(prometheus.Labels{"device": device, "type": "read"}).Set(float64(stats.bytes.read))
+		c.bytes.With(prometheus.Labels{"device": device, "type": "write"}).Set(float64(stats.bytes.write))
+		c.bytes.With(prometheus.Labels{"device": device, "type": "free"}).Set(float64(stats.bytes.free))
```

Comment:
It seems free should go into a different metric because it doesn't make sense to either average or sum together with read and write?